contrib/NeonVM/global.c

```diff
@@ -0,0 +1,541 @@
+#include <assert.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "cell.h"
+#include "global.h"
+#include "number.h"
+#include "stack.h"
+#include "string.h"
+#include "util.h"
+
+#ifdef _MSC_VER
+#ifndef strdup
+#define strdup(x) _strdup(x)
+#endif
+#endif
+
+#define STRING_BUFFER_SIZE      120
+#define PDFUNC(name, func)      { name, (void (*)(TStack *))(func) }
+
+TDispatch gfuncDispatch[] = {
+    PDFUNC("concat",                    concat),
+    PDFUNC("print",                     print),
+    PDFUNC("str",                       str),
+    PDFUNC("strb",                      strb),
+    PDFUNC("ord",                       ord),
+
+    PDFUNC("sys$exit",                  sys_exit),
+
+    PDFUNC("array__append",             array__append),
+    PDFUNC("array__concat",             array__concat),
+    PDFUNC("array__extend",             array__extend),
+    PDFUNC("array__size",               array__size),
+    PDFUNC("array__slice",              array__slice),
+    PDFUNC("array__splice",             array__splice),
+    PDFUNC("array__toBytes__number",    array__toBytes__number),
+    PDFUNC("array__toString__number",   array__toString__number),
+
+    PDFUNC("boolean__toString",         boolean__toString),
+
+    PDFUNC("bytes__decodeToString",     bytes__decodeToString),
+    PDFUNC("bytes__range",              bytes__range),
+    PDFUNC("bytes__size",               bytes__size),
+    PDFUNC("bytes__splice",             bytes__splice),
+    PDFUNC("bytes__toArray",            bytes__toArray),
+    PDFUNC("bytes__toString",           bytes__toString),
+
+    PDFUNC("number__toString",          number__toString),
+
+    PDFUNC("string__append",            string__append),
+    PDFUNC("string__toBytes",           string__toBytes),
+    PDFUNC("string__length",            string__length),
+    PDFUNC("string__splice",            string__splice),
+    PDFUNC("string__substring",         string__substring),
+
+    { 0 }
+};
+
+int global_callFunction(const char *pszFunc, TStack *stack)
+{
+    uint32_t i;
+
+    i = 0;
+    while (gfuncDispatch[i].name) {
+        if (strcmp(pszFunc, gfuncDispatch[i].name) == 0) {
+            (*gfuncDispatch[i].func)(stack);
+            return 1;
+        }
+        i++;
+    }
+    return 0;
+}
+
+void print(TStack *stack)
+{
+    const Cell *s = top(stack);
+    fwrite(s->string->data, 1, s->string->length, stdout);
+    puts("");
+    pop(stack);
+}
+
+void concat(TStack *stack)
+{
+    Cell *b = cell_fromCell(top(stack)); pop(stack);
```

Comment:
This seems to be creating a copy of a cell and then destroying it later, when you don't really need to create a copy at all. Remember cells can be quite large and it's worth trying to avoid copies if you can.

Reply:
Are you referring to this in the sense that I don't need to make a physical copy because a string is just a pointer, or from some other aspect?  In this case here, I HAVE to create the copy because I need to pop the stack, to get to the item below it.  Actually, this is a very good point.  I might have a solution for this.  I'll try and write something that avoids copies.